### PR TITLE
42 metrics home

### LINKED ISSUE #42: Find home for Metrics Calculations
At the moment, similarity metrics are computed locally and stored in a JSON file that is created and stored locally. However, model training and transfer attacks (once #40 is completed) are computed on a HPC and stored on wandb.

We should:

- Work out the best home for the similarity metrics calculation (one of the HPCs? Azure?)
  - The HPC used for model training and transfer attacks was chosen for use of GPUs. We should consider how the metrics are calculated and where the greatest costs are
- Work out where to store results: we should especially consider logging on wandb with other results

### PR DESCRIPTION
**Pull request by philswatton**
This PR resolves #42, as from here additional changes should be specific to each metric rather than general:

- breaks the metrics configs into smaller configs, each with a 'baskerville' and a 'metrics' argument. the former determines whether or not to generate a slurm script for baskerville, the latter contains the dictionary to pass to DMPair
- generate_metrics_scripts will now generate either slurm or regular bash scripts.
- the generated scripts are put into subfolders of metrics_scripts, each sharing a name with the config file for which it was generated. this allows us to run different metrics separately
- adds wandb logging for the metrics
- makes argument names across the scripts consistent (i.e. use of _path)
- updates the README
- updates the test config to be consistent with the structure used in the main configs here (i.e. everything nested under 'metrics'

Note that based on the way we've set up our use of wandb, users must run model training before they are able to generate metrics.

This PR will have merge conflicts if #53 is merged into develop first (and vice versa) due to the changes to the test configs.

Individual metrics will still need to be optimized in terms of parameter selection, and where each metric should be computed (in terms of Baskerville, etc) is still to be decided on a case by case basis.

### CODE REVIEW

File: scripts/generate_metrics_scripts.py [hunk 1]
```diff
@@ -27,36 +38,63 @@ def main(
...
+    script_names = [
+        os.path.join(scripts_path, experiment_name + "_metrics.sh")
+        for experiment_name in experiment_names
+    ]
+
+    # Are we using baskerville?
+    use_baskerville = metrics_config["baskerville"]
+
+    # If yes, prepare jinja template
+    if use_baskerville:
```
Comment (lannelin) [inline scripts/generate_metrics_scripts.py:73]: we're doing this condition check twice (now and l86, inside loop). It also defines new variables that are then only used later after condition
I suggest restructure to two separate methods from this point:
`write_baskerville_scripts` and `write_default_scripts`, or similar. Each of these methods would loop through the combos and do the writes

File: tests/testconfig/metrics.yaml
Comment (lannelin) [inline tests/testconfig/metrics.yaml:1]: this is now different to the way we structure metrics in reality. Is this going to cause any issues?

**Review by lannelin — COMMENTED** [review]
looks good, haven't tested on Baskerville. I think this will be affected by other outstanding PR so let's hold on merge as discussed.

File: tests/testconfig/metrics.yaml
Comment (philswatton) [reply tests/testconfig/metrics.yaml:1]: Following stand-up discussion they do share structure as the 'real' yaml files still (can) contain multiple metrics in the same config

File: tests/test_data.py [hunk 2]
```diff
@@ -401,7 +401,7 @@ def test_compute_similarity_error():
 
     # Read in the configs
     with open(metrics_config_path, "r") as stream:
-        mmd_config = yaml.safe_load(stream)
+        mmd_config = yaml.safe_load(stream)["metrics"]
```
[the remark this answers is not in the document]
Comment (philswatton) [reply tests/test_data.py:404]: done

File: scripts/generate_metrics_scripts.py (hunk 1 above)
Comment (philswatton) [reply scripts/generate_metrics_scripts.py:77]: Have factored these out into utils, and called the Baskerville one in model training and attacking too

**Comment by philswatton** [comment]
I've pushed some changes to the PR. The biggest one by far is factoring out the script generation into utils, and doing this across the 3 generation scripts.

In terms of testing on Baskerville, scripts generate correctly and model training/attacks seem to run fine. I haven't yet run the metrics scripts on Baskerville because we don't yet have a metric with its hyperparameters settled that we're ready to log the results of.

**Review by joannacknight — APPROVED** [review]
Only a few comments from me on the README file.

File: scripts/README.md [hunk 3]
```diff
@@ -2,61 +2,81 @@
...
+where `"drop-only"` is an example of an experiment group with a named config and `"mmd"` is an example of a named metric config (see below for more information on configs). It will generate bash scripts in `ROOT/metrics_scripts/x`, where `x` is also the name of the metric config used. Depending on the `baskerville` entry in the metric config, this will generate either regular bash scripts or slurm scripts. Note additional optional arguments that can be passed to the generation script include `--dataset_config_path` and `--metrics_config_path`.
 
-Once this is done, you can run all of the generated scripts for a given experiment group by calling `scripts/train_all.sh` followed by the name of the experiment group, e.g.
+Once this is done, you can run all of the generated scripts for a given experiment group and metric config by calling either `scripts/slurm_metric_all.sh` where slurm scripts have been generated or `scripts/shell_metric_all.sh` followed by the name of the metric config and the name of the experiment group, e.g.
 
 ```bash
-scripts/train_all.sh drop-only
+scripts/slurm_metric_all.sh mmd drop-only
```
[the remark this answers is not in the document]
Comment (philswatton) [reply scripts/README.md]: good spot!

**Pull request merged**